src/Std/System/Uri.lean

```diff
@@ -0,0 +1,109 @@
+/-
+Copyright (c) 2019 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Lovett
+-/
+
+namespace System
+namespace Uri
+namespace UriEscape
+
+def hexDigitToNat (c : Char) : Option Nat :=
+  if '0' ≤ c ∧ c ≤ '9' then some (c.toNat - '0'.toNat)
+  else if 'a' ≤ c ∧ c ≤ 'f' then some (c.toNat - 'a'.toNat + 10)
+  else if 'A' ≤ c ∧ c ≤ 'F' then some (c.toNat - 'A'.toNat + 10)
+  else none
+
+def decodeUri (uri : String) : String := Id.run do
+  let mut decoded : ByteArray := ByteArray.empty
+  let len := uri.utf8ByteSize
+  let mut pos : Nat := 0
+  let mut i := uri.mkIterator
+  while pos < len do
+    let c := i.curr
+    if c == '%' && pos + 2 < len then
```

Comment:
I think this check might fail if you try just the string `"%%"`.

Reply:
Oh, good catch, fixed, thanks.